### PR TITLE
Enhancement/noaa taf

### PR DESCRIPTION
**Pull request by IamLuisAUT**
- Rename TafService to CheckWxService
- Create NoaaService

### CODE REVIEW

File: src/Service/Taf/NoaaService.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * @var \Cake\Console\ConsoleIo|null
+     */
+    protected ?ConsoleIo $_io;
+
+    /**
+     * @var string|null
+     */
+    protected ?string $_rawTaf = null;
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: We can remove this property.

File: src/Service/Taf/NoaaService.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    public function getTaf(): void
+    {
+        $airports = $this->Airports->find()
+            ->select(['icao'])
+            ->enableHydration(false)
+            ->all()
+            ->extract('icao')
+            ->toArray();
+
+        $rawTaf = $this->_fetchTaf($airports);
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: Minor thing, but we could spare the request if we have no airports (which is unlikely ofc 🙂)

File: src/Service/Taf/NoaaService.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        if ($rawTaf['data']['@attributes']['num_results'] > 1) {
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: We have to handle an empty array.

File: src/Service/Taf/NoaaService.php [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+ */
+class NoaaService
+{
+    use ModelAwareTrait;
+    use ZMQContextTrait;
+
+    /**
+     * @var string
+     */
+    protected const _NOAA_URL = 'https://aviationweather.gov/adds/dataserver_current/httpparam?dataSource=tafs&requestType=retrieve&format=xml&mostRecentForEachStation=true&hoursBeforeNow=0&stationString=';
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: Set it to `https://aviationweather.gov/adds/dataserver_current/httpparam`.
See the comment on `NoaaService::_fetchTaf`

File: src/Service/Taf/NoaaService.php [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        $this->_io = $io;
+    }
+
+    /**
+     * @param array $tafStations TAF stations
+     * @return array
+     */
+    protected function _fetchTaf(array $tafStations): array
+    {
+        $response = $this->_client->get(self::_NOAA_URL . implode('%20', $tafStations), [], [
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: What about something like this?

```
$response = $this->_client->get(self::_NOAA_URL, [
    'dataSource' => 'tafs',
    'requestType' => 'retrieve',
    'format' => 'xml',
    'mostRecentForEachStation' => 'true',
    'hoursBeforeNow' => 0,
    'stationString' => urlencode(implode(',', $tafStations)),
], [
    'type' => 'xml',
]);
```

File: src/Service/Taf/NoaaService.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        $this->loadModel('Airports');
+        $this->loadModel('Taf');
+
+        $this->_client = new Client();
+    }
+
+    /**
+     * @return void
+     */
+    public function getTaf(): void
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php:42]: Maybe add a `@FIXME`, as we most likely need to make more than one request once the overall number of airports grows in the DB. Is there a limit on the `stationString` parameter?

File: src/Service/Taf/NoaaService.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    protected function _fetchTaf(array $tafStations): array
+    {
+        $response = $this->_client->get(self::_NOAA_URL . implode('%20', $tafStations), [], [
+            'headers' => [
+            ],
+        ]);
+        if ($response->isOk()) {
+            $xmlTaf = $response->getXml();
+
+            return json_decode(json_encode($xmlTaf), true);
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: We should handle eventual errors.

File: src/Service/Taf/NoaaService.php (hunk 6 above)
Comment (IamLuisAUT) [reply src/Service/Taf/NoaaService.php:42]: denied request for approx 950 airports. request with 900 given airports was approved -> 1000 refused. @FIXME still not a bad idea

File: src/Service/Taf/NoaaService.php [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+            ->select(['icao'])
+            ->enableHydration(false)
+            ->all()
+            ->extract('icao')
+            ->toArray();
+
+        if($airports) {
+            $rawTaf = $this->_fetchTaf($airports);
+    
+            if($rawTaf && $rawTaf['data']['@attributes']['num_results'] && $rawTaf['data']['TAF']) {
```
Comment (cleptric) [inline src/Service/Taf/NoaaService.php]: We should check if the array actually contains values. Something like `!empty()` or `isset()` depending on how the API returns it's data.

**Review by cleptric — APPROVED** [review]
Thanks Luis!

**Pull request merged**